Disable ssp on -nostdlib, -nodefaultlibs and -ffreestanding 
Change the buffer size.

--- a/gcc/gcc.c	2017-07-04 09:15:57.740793000 +0200
+++ b/gcc/gcc.c	2018-03-02 13:58:44.387741114 +0100
@@ -977,6 +977,12 @@
 #define LINK_GCC_C_SEQUENCE_SPEC "%G %{!nolibc:%L %G}"
 #endif
 
+#ifdef ENABLE_DEFAULT_SSP
+#define NO_SSP_SPEC "%{nostdlib|nodefaultlibs|ffreestanding:-fno-stack-protector} "
+#else
+#define NO_SSP_SPEC ""
+#endif
+
 #ifndef LINK_SSP_SPEC
 #ifdef TARGET_LIBC_PROVIDES_SSP
 #define LINK_SSP_SPEC "%{fstack-protector|fstack-protector-all" \
@@ -1273,7 +1279,7 @@
  %{-version:--version}\
  %{-help=*:--help=%*}\
  %{!fsyntax-only:%{S:%W{o*}%{!o*:-o %w%b.s}}}\
- %{fsyntax-only:-o %j} %{-param*}\
+ %{fsyntax-only:-o %j} %{-param*} " NO_SSP_SPEC "\
  %{coverage:-fprofile-arcs -ftest-coverage}\
  %{fprofile-arcs|fprofile-generate*|coverage:\
    %{!fprofile-update=single:\
--- a/gcc/params.opt	2020-05-07 18:49:59.000000000 +0800
+++ b/gcc/params.opt	2020-05-07 18:49:59.000000000 +0800
@@ -853,7 +853,7 @@
 The maximum number of SSA_NAME assignments to follow in determining a value.
 
 -param=ssp-buffer-size=
-Common Joined UInteger Var(param_ssp_buffer_size) Init(8) IntegerRange(1, 65536) Param Optimization
+Common Joined UInteger Var(param_ssp_buffer_size) Init(4) IntegerRange(1, 65536) Param Optimization
 The lower bound for a buffer to be considered for stack smashing protection.
 
 -param=stack-clash-protection-guard-size=
